src/render/opengles2/SDL_render_gles2.c

```diff
@@ -955,6 +955,11 @@ SetDrawState(GLES2_RenderData *data, const SDL_RenderCommand *cmd, const GLES2_I
             }
 #endif
             data->glBindTexture(tdata->texture_type, tdata->texture);
+
+            /* Generate mipmaps for minification if scaling using mipmaps is enabled */
+            if(texture->scaleMode == SDL_ScaleModeBest && tdata->texture_type == GL_TEXTURE_2D) {
+                data->glGenerateMipmap(tdata->texture_type);
```

Comment:
Re-generating mipmaps before every draw would add a hefty performance penalty. This codepath also only gets run if the last drawn texture isn't this texture, so there's a bug here in a situation like RenderCopy -> update texture -> RenderCopy when the same texture is used for all 3 operations.

Reply:
> Mipmaps have various tradeoffs (improved runtime performance versus increased memory size, aesthetic choice, (lack of) mipmap LOD bias, etc). Same with trilinear and anisotropic filtering versus bilinear.
> 
> Typically graphics APIs will expose an option on texture creation for whether to use mipmaps or not, as well as extra filter options per sampler/texture in addition to bilinear and point/nearest-neighbour filtering. I think doing that with SDL would be a better approach than retrofitting new and surprising features into APIs people already use in shipping code.

That is what the SDL_ScaleMode* enum is for is it not? SDL currently does not provide us with an option on texture creation itself and we have to rely on the global SDL_ScaleMode* hint on how to handle mipmaps on texture creation.

Currently we have:
```
    SDL_ScaleModeNearest, /**< nearest pixel sampling */
    SDL_ScaleModeLinear,  /**< linear filtering */
    SDL_ScaleModeBest     /**< anisotropic filtering */
    
```
 where _SDL_ScaleModeBest_ does nothing on the OpenGL backend. Adding for example a new enum value called _SDL_ScaleModeMipmap_ would prevent changes any behaviour for existing shipped code while allowing the API user to have access to this new feature. But then the enum value _SDL_ScaleModeMipmap_ would only have an effect when using OpenGL.

Reply:
>  SDL currently does not provide us with an option on texture creation itself and we have to rely on the global SDL_ScaleMode* hint on how to handle mipmaps on texture creation.

That's what your pull request does, yeah. But I think it's better to add new APIs for new features instead of retrofitting them into old APIs. Especially in the case of mipmaps, where using them has implications that aren't strictly "this is always better" and where they actually do need to be specified on texture creation in other graphics APIs.

Reply:
> Re-generating mipmaps before every draw would add a hefty performance penalty. This codepath also only gets run if the last drawn texture isn't this texture, so there's a bug here in a situation like RenderCopy -> update texture -> RenderCopy when the same texture is used for all 3 operations.

Yes this is a bug. The mipmaps should only be generated once right after SDL_CreateTexture

Reply:
> The mipmaps should only be generated once right after SDL_CreateTexture

That would introduce more bugs unless you make sure mipmaps are regenerated after updating or unmapping a texture, as well as after switching the active render target away from a mipmapped texture.

Reply:
* Mipmaps should now be generated on texture creation, whenever changing the scaling method and when the texture is updated.
* Mipmap generation/filtering is now decoupled from scaling method (nearest vs linear) and is set per texture defaulted from a global renderer hint (applicable to OpenGL).
* Added support for OpenGL ES 1.0 as well

As I understand it if the active render target is a texture which has mipmaps generated, then we need to re-generate mipmaps as well.
Can you explain what you mean by "unmapping a texture"? When does this happen?

Reply:
Oh my bad, I meant UnlockTexture. However it looks like the gles2 UnlockTexture implementation just calls UpdateTexture (whereas some other backends have different code), so it should be fine in that renderer.